### PR TITLE
Attempt sending Duo "push" response automatically

### PR DESCRIPTION
**Pull request by rca**
99% of the time, the Duo challenge response is going to be `push` when connecting to the VPN, and the code should try this automatically.

This patch sends the push response automatically and will fallback to prompting the user when that fails.

### CODE REVIEW

**Comment by PaulHatch** [comment]
Tried this out and it works great, would be a nice addition here.

**Comment by Magikarplvl4** [comment]
Any chance this PR will be merged @xdesai @ben-duo ?

**Comment by xdesai** [comment]
Thanks for raising this request. Sorry that it sat for so long with no word from us!
I'll need to take this back and run it by the team to see if this is something we're interested in doing.

After a first glance I can tell you now that we'll probably need a couple of changes.
1. We'll want to create a constant at the top of the file called AUTOPUSH. This is essentially the configuration since we don't really have one for this integration. And following our opt-in approach we'll probably set it to false by default.

2. For our other integrations we don't typically have autopush and also allow users to pick an auth factor. I think for UX reasons we may just choose autopush only instead of prompting as a fallback.

Before you spend more time updating though I'll check with our product team so we can decide if this is something we have time to invest into.
Thanks again for the PR and the idea!

**Comment by rca** [comment]
Thanks for the feedback. I’ll wait to hear back.

**Comment by xdesai** [comment]
Alright, just finished talking with our product team.
If you would be able to address the feedback that I mentioned above we'll set aside some time in the next few weeks to review the changes, test them out locally, and hopefully get these merged in. Just for transparency sake we don't have time in our current development cycle so it will be at least 2 weeks before we can look at any changes you make.
Thanks for your work on this so far!
Xander

**Comment by Magikarplvl4** [comment]
Hi @xdesai,

Thanks for looking into this and coming back to us with an update.
I much appreciate it!

**Comment by rca** [comment]
Changes made to use the `AUTOPUSH` constant on the top of the file for setting the behavior.

I have verified that setting it to `True` works as expected.   

I am, however, having trouble with my VPN client showing me a prompt to enter the challenge response.  If you can verify `AUTOPUSH = False` works for you, that would be great.  I am waiting on a response from my vpn client's vendor to see if it's a bug on the client side.

Thanks.

**Comment by rca** [comment]
@xdesai can you please confirm you see my updates and that they address your requests.

Thank you.

**Comment by xdesai** [comment]
I took an early look over this and I realized I totally forgot to guide you towards our particular way of doing autopush.
The way we do this in other integrations like Duo Unix and the Duo Authentication Proxy is to actually use the user's default factor.
What you have coded now always sets the factor to "push". While this will work for some users, it would be really nice to be able to "autopush" users who may only have a landline. Or a user who has two phones and wants the pushes to go to a certain phone.

So, how do we get the default factor that the user has configured? The answer is to use the preauth call.
https://duo.com/docs/authapi-v1#/preauth
This call returns a list of configured factors and one of them is marked as "default". What we'll want to do is save the default factor from the preauth call and then pass it in as the factor for the auth call.

The bummer is that this is going to require a slightly larger change than you have today. This is because the preauth function in this repo doesn't pass back the factors. So we'll need to change the `def preauth` function to return a list of factors or at least return the default factor.
Maybe something like this

```
response = self.json_api_call('POST', '/rest/v1/preauth', params)
result = response.get('result')
factors = response.get('factors', None)
```
and then all the return statements in the `def preauth function`
can return
```
return (result, factors, msg)
```
instead of just
```
return (result, msg)
```


Then in `post_auth_cr`
```
result, factors, msg = api.preauth(username, ipaddr)
if result == API_RESULT_AUTH:
    if AUTOPUSH:
            autopush_factor = factors["default"]
            log("try autopush")
            crstate["autopush_factor"] = autopush_factor
            post_auth_cr(authcred, attributes, authret, info, crstate)
```

I know this is a decent amount of feedback, but does that make sense? Thanks again for working with us on this!

**Comment by rca** [comment]
@xdesai I made a couple of small tweaks to your suggestion:

- `autopush_factor = factors["default"]` can potentially fail if `factors = response.get('factors', None)` is `None`, so I updated the default factor to be an empty dict and then use `.get('default')` to get the value
- rather than returning yet another item in a tuple, i just return the entire response wrapped in a custom class that provides helpers to get the desired attributes.  this makes it easier to get other attributes in the future and alleviates some potential confusion over the order of the items in the tuple.

I believe the patch addresses all your concerns.  Please feel free to make any additional tweaks necessary on this PR.

Thank you!

**Comment by xdesai** [comment]
Thanks for making those updates!
I really like the preauth response class that you created. I'm making a couple small tweaks on my end and doing some testing. I'll push up my changes soon so you can review them as well.

**Comment by xdesai** [comment]
Hey @rca I was gonna push my changes up for you to review and test, but I got a permission denied. Can you make sure your pull request allows edits from maintainers? Thanks!
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by rca** [comment]
It was already checked:

image

**Comment by xdesai** [comment]
Oh weird! I must've done something wrong on my end. Thanks for confirming! I'll keep poking around on my end and see if I can figure out the right command to use.

**Comment by xdesai** [comment]
Okay got the update to go through.
@rca if you could pull down my changes and test it with your local environment that would be really appreciated!
I'll have someone from my team look this over, we'll make any final edits based on their feedback and then hopefully get this merged soon.

File: duo_openvpn_as.py
Comment (rca) [inline duo_openvpn_as.py:568]: This does not appear to be used anywhere.

**Review by rca — COMMENTED** [review]
My only comment is that the status setter is not used anywhere.  Otherwise, LGTM.  Thanks!

File: duo_openvpn_as.py
Comment (xdesai) [reply duo_openvpn_as.py:568]: ah good point. I just included it for completeness since the "alias" `msg` can set, I figured allowing one to set the `status` made sense.
Thanks for looking this over!
If you get a chance to pull these changes and test on your local OpenVPN server let me know. I was thinking of waiting to merge until I get the thumbs up from you. But let me know if you don't have capacity to run that test right now.

Comment (rca) [reply duo_openvpn_as.py:568]: When I saw the setters initially I was a bit confused because adding setters lets you modify an upstream response, which should technically be immutable.  Then I saw why you added the message setter because I missed a use-case in the code I modified, so I get it, but I sort of feel that changing the upstream response shouldn't be encouraged.

That said, I don't think it's a big deal.  👍🏽

**Comment by rca** [comment]
> Okay got the update to go through.
> @rca if you could pull down my changes and test it with your local environment that would be really appreciated!
> I'll have someone from my team look this over, we'll make any final edits based on their feedback and then hopefully get this merged soon.

Yes, please definitely check on your end.  It might take a while to test on my end; I only have a live VPN server and changing this script could potentially disconnect others, so I have to wait for a maintenance window.

**Comment by rca** [comment]
I am confirming that this version of the Duo post-auth script is working fine for me 👍 

I just made a small change to the autopush log where I also log the `autopush_factor` that was selected.  Additionally, I tweaked the string interpolation to use the old `%` formatting, which I found to be consistent with the rest of the code and compatible with OpenVPN AS; for some reason the other interpolation formats didn't seem to work, so I stayed away from them.

Thanks.

**Review by xdesai — APPROVED** [review]
We have to be very careful logging the "factor" because sometimes the factor can be a passcode. That was part of the reason I originally left it out. However, after looking into it further we actually know for certain that the autopush factor will never be a passcode so it is safe to log in this scenario.

**Comment by xdesai** [comment]
Thanks for all your work on this! It was good to work with you and I'm happy we were able to get this merged in.
Happy autopushing! :)

**Pull request merged**
